### PR TITLE
middleware: ignore iat claim in jwt

### PR DESCRIPTION
**Pull request by raghavendra-talur**
### What does this PR achieve? Why do we need it?
This PR implements new claims type and validation function. We need it due to inability of jwt library to handle clock skew.

### Notes for the reviewer
1. required_claims is embedded in validation function and missing claims error is replaced with unauthorized
2. As claims type has changed from map to struct, claims are now accessed as members and not indexed.

### CODE REVIEW

**Comment by phlogistonjohn** [comment]
Before I look at the code, a thought: Should we make this a configurable behavior? Would there be users of heketi out there who would prefer to suffer from the "token used before issued" errors at the expense of having potentially improved security?

File: middleware/jwt_test.go [hunk 1]
```diff
@@ -394,6 +396,77 @@ func TestJwt(t *testing.T) {
 	tests.Assert(t, r.StatusCode == http.StatusOK)
 	tests.Assert(t, called == true)
 }
+func TestJwtIgnoreIAT(t *testing.T) {
```
Comment (phlogistonjohn) [inline middleware/jwt_test.go]: I guess gofmt doesn't care about this, but I do: there should be an empty line between the closing curly brace of the previous function and the start of the new one.

File: middleware/jwt.go [hunk 2]
```diff
@@ -25,12 +26,44 @@ import (
...
+	vErr := new(jwt.ValidationError)
+	now := jwt.TimeFunc().Unix()
+
+	if c.VerifyExpiresAt(now, true) == false {
+		delta := time.Unix(now, 0).Sub(time.Unix(c.ExpiresAt, 0))
+		vErr.Inner = fmt.Errorf("Token is expired by %v", delta)
+		vErr.Errors |= jwt.ValidationErrorExpired
+	}
+
+	// We skip the check for "iat"
```
Comment (phlogistonjohn) [inline middleware/jwt.go]: We should either eliminate, modify, or conditionalize this block but we should not leave it just as a comment.

**Comment by raghavendra-talur** [comment]
@phlogistonjohn Changes done as requested. Also, I found a cool way to test ENV based configuration :  https://medium.com/@povilasve/go-advanced-tips-tricks-a872503ac859. Check new tests added.

File: middleware/jwt.go [hunk 3]
```diff
@@ -15,22 +15,76 @@ import (
...
+	"github.com/heketi/heketi/pkg/utils"
 )
 
+var (
+	logger          = utils.NewLogger("[jwt]", utils.LEVEL_DEBUG)
+	iatLeeway int64 = 120
+)
+
+func init() {
+	if leeway, exists := os.LookupEnv("HEKETI_JWT_IAT_LEEWAY_SECONDS"); exists {
```
Comment (Madhu-1) [inline middleware/jwt.go]: I have a question, what happens if we set ```HEKETI_JWT_IAT_LEEWAY_SECONDS``` as a negative value.

File: middleware/jwt.go [hunk 4]
```diff
@@ -15,22 +15,76 @@ import (
...
+		vErr.Inner = fmt.Errorf("Token is expired by %v", delta)
+		vErr.Errors |= jwt.ValidationErrorExpired
+		logger.LogError("exp validation failed: %v", vErr.Error())
+	}
+
+	// "iat" check
+	if now < c.IssuedAt-iatLeeway {
+		vErr.Inner = fmt.Errorf("Token used before issued")
+		vErr.Errors |= jwt.ValidationErrorIssuedAt
+		logger.LogError("iat validation failed: %v, time now: %v, time issued: %v", vErr.Error(), now, c.IssuedAt)
```
Comment (Madhu-1) [inline middleware/jwt.go]: ```IssuedAt``` and ```now``` are both in Unix form.
can we convert back the Unix time to UTC time for more readability on logging

**Comment by phlogistonjohn** [comment]
retest this please

File: middleware/jwt.go (hunk 4 above)
Comment (raghavendra-talur) [reply middleware/jwt.go]: done

File: middleware/jwt.go (hunk 3 above)
Comment (raghavendra-talur) [reply middleware/jwt.go]: Handled now and added a test.

File: middleware/jwt_test.go [hunk 5]
```diff
@@ -391,10 +390,414 @@ func TestJwt(t *testing.T) {
 	req.Header.Set("Authorization", "bearer "+tokenString)
 	r, err := http.DefaultClient.Do(req)
 	tests.Assert(t, err == nil)
-	tests.Assert(t, r.StatusCode == http.StatusOK)
+	tests.Assert(t, r.StatusCode == http.StatusOK, r.StatusCode, r.Status)
+	tests.Assert(t, called == true)
+}
+func TestJwtLeewayIAT(t *testing.T) {
```
Comment (phlogistonjohn) [inline middleware/jwt_test.go]: Missing blank line between closing curly and new function?

**Pull request merged**